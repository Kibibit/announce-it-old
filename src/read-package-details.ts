import fs from 'fs-extra';
import { get, isString } from 'lodash';
import { IBranchObject } from './announce-it';

export interface IPackageDetails {
  name: string;
  version: string;
  description: string;
  author: string;
  homepage: string;
  repository: {
    type: string;
    url: string;
  };
  announcements: {
    tweet: string;
    includeUnstable?: boolean;
  };
  release: {
    branches: Array<string | IBranchObject>;
  }
}

export function readPackageDetails(root: string): Promise<IPackageDetails> {
  return fs.readJson(`${ root }/package.json`)
    .then((packageDetails: Partial<IPackageDetails>) => {
      if (!isString(get(packageDetails, 'announcements.tweet'))) {
        throw new Error('no tweet template found. please see the readme for more details');
      }

      // here, we KNOW the package has the announcement template string
      return packageDetails as IPackageDetails;
    });
}
