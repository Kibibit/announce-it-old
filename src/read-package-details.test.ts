import fs from 'fs-extra';
import { assign } from 'lodash';

import { IPackageDetails, readPackageDetails } from './read-package-details';

const packageDetails: IPackageDetails = {
  name: 'package-name',
  description: 'package description',
  version: '1.0.0',
  author: 'test@test.com',
  homepage: 'pizza.com',
  repository: {
    type: '',
    url: ''
  },
  announcements: {
    tweet: 'nice!'
  }
};

jest.mock('fs-extra');

describe('readPackageDetails', () => {
  it('should return the package details', () => {
    // @ts-ignore
    fs.readJson.mockResolvedValue(packageDetails);
    return expect(readPackageDetails('rootFolder')).resolves.toMatchSnapshot();
  });

  it('should throw error if reading package.json failed', () => {
    // @ts-ignore
    fs.readJson.mockRejectedValue(new Error('failed reading json file'));
    return expect(readPackageDetails('rootFolder')).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should throw error if missing tweet from package.json', () => {
    const missingDetails: Partial<IPackageDetails> = assign({}, packageDetails);
    missingDetails.announcements = undefined;
    // @ts-ignore
    fs.readJson.mockResolvedValue(missingDetails);

    return expect(readPackageDetails('rootFolder')).rejects.toThrowErrorMatchingSnapshot();
  });
});
