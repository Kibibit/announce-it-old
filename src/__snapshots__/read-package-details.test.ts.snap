// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readPackageDetails should return the package details 1`] = `
Object {
  "announcements": Object {
    "tweet": "nice!",
  },
  "author": "test@test.com",
  "description": "package description",
  "homepage": "pizza.com",
  "name": "package-name",
  "release": Object {
    "branches": Array [
      "TEST",
    ],
  },
  "repository": Object {
    "type": "",
    "url": "",
  },
  "version": "1.0.0",
}
`;

exports[`readPackageDetails should throw error if missing tweet from package.json 1`] = `"no tweet template found. please see the readme for more details"`;

exports[`readPackageDetails should throw error if reading package.json failed 1`] = `"failed reading json file"`;
